### PR TITLE
#79: Add Sphinx error propogation

### PR DESCRIPTION
**Pull request by AndrewSamokhvalov**
Issue: #79

### CODE REVIEW

**Comment by AndrewSamokhvalov** [comment]
Depends on [this](https://github.com/lightningnetwork/lightning-onion/pull/8) issues

File: htlcswitch/switch.go [hunk 1]
```diff
@@ -337,54 +345,59 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
 			))
-			err := errors.Errorf("unable to find link with "+
+			err = errors.Errorf("unable to find link with "+
 				"destination %v", packet.dest)
 			log.Error(err)
 			return err
 		}
-		interfaceLinks, _ := s.getLinks(targetLink.Peer().PubKey())
-
-		// Try to find destination channel link with appropriate
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Why was this removed? All links should be checked for their available bandwidth.

File: htlcswitch/switch.go [hunk 2]
```diff
@@ -337,54 +345,59 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
 
 		// If the channel link we're attempting to forward the update
 		// over has insufficient capacity, then we'll cancel the htlc
 		// as the payment cannot succeed.
-		if destination == nil {
-			// If packet was forwarded from another channel link
-			// than we should notify this link that some error
-			// occurred.
-			reason := []byte{byte(lnwire.InsufficientCapacity)}
+		if destination.Bandwidth() < htlc.Amount {
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: An error other than `FailInsufficientCapacity` should be sent.

File: htlcswitch/switch.go [hunk 3]
```diff
@@ -398,17 +411,25 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 			source.ShortChanID(),
 			destination.ShortChanID(),
 			htlc.PaymentHash,
+			packet.obfuscator,
 		)); err != nil {
-			reason := []byte{byte(lnwire.UnknownError)}
+			failure := lnwire.FailPermanentChannelFailure{}
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Not sure this should send a permanent failure, as the error may have been transient.

File: htlcswitch/switch.go [hunk 4]
```diff
@@ -433,6 +454,11 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 			return err
 		}
 
+		// If this is failure than we need to obfuscate the error.
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: It seems that a case above for the malformed onion error needs to be added. Otherwise, where is this handled currently in the PR? (still getting through it, so may be in the next few commits after this one).

File: htlcswitch/test_utils.go [hunk 5]
```diff
@@ -26,8 +30,42 @@ var (
...
+		0xd, 0xe7, 0x95, 0xe4, 0xb7, 0x25, 0xb8, 0x4d,
+		0x1e, 0xb, 0x4c, 0xfd, 0x9e, 0xc5, 0x8c, 0xe9,
+	}
+
+	_, testPubKey = btcec.PrivKeyFromBytes(btcec.S256(), testPrivKey)
+	testSig       = &btcec.Signature{
+		R: new(big.Int),
+		S: new(big.Int),
+	}
+	_, _ = testSig.R.SetString("63724406601629180062774974542967536251589935445068131219452686511677818569431", 10)
```
Comment (Roasbeef) [inline htlcswitch/test_utils.go]: Formatting is messed up here. Also, the method is called, but the values aren't captured anywhere. Seems to be left over from a copy paste somewhere.

File: routing/router.go [hunk 6]
```diff
@@ -111,8 +114,8 @@ type Config struct {
 	// forward a fully encoded payment to the first hop in the route
 	// denoted by its public key. A non-nil error is to be returned if the
 	// payment was unsuccessful.
-	SendToSwitch func(firstHop *btcec.PublicKey,
-		htlcAdd *lnwire.UpdateAddHTLC) ([32]byte, error)
+	SendToSwitch func(firstHop *btcec.PublicKey, htlcAdd *lnwire.UpdateAddHTLC,
```
Comment (Roasbeef) [inline routing/router.go]: Layer violation. The router shouldn't have to know about the interfaces of the switch, or that encrypted errors are even used at all. 

Instead something like the following should be added as a parameter:
```go
type Circuit struct {
    SessionKey *btcec.PrivateKey

    Path []*btcec.PublicKey
}
```

Such a struct should then be added to the `sphinx` package. In the server, when creating the `ChannelRouter`, the closure there should then morph the `sphinx.Circuit` instance into the `htlcswitch.Deobfuscator`. A little indirection goes a long way.

File: routing/router.go [hunk 7]
```diff
@@ -904,7 +908,16 @@ func generateSphinxPacket(route *Route, paymentHash []byte) ([]byte, error) {
 		}),
 	)
 
-	return onionBlob.Bytes(), nil
+	// Initialize the onion error deobfuscator in order to be able to decode
```
Comment (Roasbeef) [inline routing/router.go]: In line with the comment above, a `sphinx.Circuit` should be created and returned here instead.

File: routing/router.go [hunk 8]
```diff
@@ -847,11 +850,12 @@ func (r *ChannelRouter) FindRoutes(target *btcec.PublicKey, amt btcutil.Amount)
 	return validRoutes, nil
 }
 
-// generateSphinxPacket generates then encodes a sphinx packet which encodes
+// generateOnionBlob generates then encodes a sphinx packet which encodes
```
Comment (Roasbeef) [inline routing/router.go]: Why the rename? Diffs should be kept as minimal as possible. Please revert. Thanks!

File: peer.go [hunk 9]
```diff
@@ -276,6 +276,33 @@ func (p *peer) Start() error {
 	return nil
 }
 
+// getEdgeInfo returns the channel edge info.
+// TODO(andrew.shvv) place it inside the router after it will be clear that
+// this method is needed inside the channel link.
+func (p *peer) getEdgeInfo(chanID lnwire.ChannelID) (*channeldb.ChannelEdgePolicy,
```
Comment (Roasbeef) [inline peer.go]: Why does this method need to reside as part of the peer struct? It seems that instead a closure should be created at the site of the initialization of the configuration for a channel link.

File: peer.go [hunk 10]
```diff
@@ -276,6 +276,33 @@ func (p *peer) Start() error {
 	return nil
 }
 
+// getEdgeInfo returns the channel edge info.
```
Comment (Roasbeef) [inline peer.go]: The commit title needs to be reworked a bit, how about: `lnd: hook up encrypted onion error scaffolding`.

File: server.go [hunk 11]
```diff
@@ -71,7 +71,7 @@ type server struct {
 
 	utxoNursery *utxoNursery
 
-	sphinx *sphinx.Router
+	onionProcessor *htlcswitch.OnionProcessor
```
Comment (Roasbeef) [inline server.go]: Prior name didn't need to be changed IMO.

File: htlcswitch/switch.go [hunk 12]
```diff
@@ -88,6 +88,10 @@ type Config struct {
 	// or forced unilateral closure of the channel initiated by a local
 	// subsystem.
 	LocalChannelClose func(pubKey []byte, request *ChanClose)
+
+	// UpdateTopology sends the onion error failure topology update to router
+	// subsystem.
+	UpdateTopology func(msg *lnwire.ChannelUpdate)
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Perhaps an error should be returned, if only for logging purposes.

File: discovery/service.go [hunk 13]
```diff
@@ -19,7 +19,7 @@ import (
 // networkMsg couples a routing related wire message with the peer that
 // originally sent it.
 type networkMsg struct {
-	peer *btcec.PublicKey
+	peer []byte
```
Comment (AndrewSamokhvalov) [inline discovery/service.go]: Good point, sometimes I have to stop coding after `00:00`...

File: lnwire/onion_error.go [hunk 14]
```diff
@@ -40,6 +40,8 @@ const (
 // unaltered to the source of the HTLC within the route.
 type FailCode uint16
 
+// The currently defined onion failue types within this current version of the
```
Comment (AndrewSamokhvalov) [inline lnwire/onion_error.go]: I believe that ordering is wrong only in `github` commit viewer.

File: lnwire/onion_error.go [hunk 15]
```diff
@@ -0,0 +1,879 @@
...
+	FlagUpdate FailCode = 0x1000
+)
+
+// FailCode specifies the precise reason that an upstream HTLC was cancelled.
+// Each UpdateFailHTLC message carries a FailCode which is to be passed back
+// unaltered to the source of the HTLC within the route.
+type FailCode uint16
+
+const (
+	CodeInvalidRealm                  FailCode = FlagBadOnion | 1
```
Comment (AndrewSamokhvalov) [inline lnwire/onion_error.go]: I have added only one comment on top of the `const` block, as you did it with `lnwire` messages types.

File: lnwire/onion_error.go [hunk 16]
```diff
@@ -0,0 +1,879 @@
...
+
+// Code returns the failure unique code.
+// NOTE: Part of the Failure interface.
+func (f FailInvalidRealm) Code() FailCode {
+	return CodeInvalidRealm
+}
+
+// FailTemporaryNodeFailure is returned if an otherwise unspecified transient
+// error occurs for the entire node.
+// NOTE: might be returned by any node.
```
Comment (AndrewSamokhvalov) [inline lnwire/onion_error.go]: I think that keeping it as a distinct note, might reduce time for programmer to better understand what expect from this error, without reading full description.

File: routing/router.go (hunk 6 above)
Comment (AndrewSamokhvalov) [inline routing/router.go]: I already have such struct:
```
// OnionDeobfuscator represents the serializable object which encapsulate the
// all necessary data to properly de-obfuscate previously obfuscated data.
// In context of Lightning Network the data which have to be deobfuscated
// usually is onion failure.
type OnionDeobfuscator struct {
	sessionKey  *btcec.PrivateKey
	paymentPath []*btcec.PublicKey
}
```

Comment (AndrewSamokhvalov) [inline routing/router.go]: I moved the creation of the `htlcswitch.Deobfuscator` to the `SendToSwitch` handler, now it takes the `sphinx.OnionDeobfuscator` as input, and now seems like layer violation have been avoided, at least in terms of importing the `htlcswitch` package.

File: htlcswitch/packet.go [hunk 17]
```diff
@@ -29,12 +29,27 @@ type htlcPacket struct {
...
 
 	// amount is the value of the HTLC that is being created or modified.
-	//
 	// TODO(andrew.shvv) should be removed after introducing sphinx payment.
 	amount btcutil.Amount
 
 	// htlc lnwire message type of which depends on switch request type.
 	htlc lnwire.Message
+
+	// obfuscator is entity which is needed to make the obfuscation of the
```
Comment (AndrewSamokhvalov) [inline htlcswitch/packet.go]: 1. As I recall we agreed on storing the `obfuscators` in the switch payment circuits, because in the next `PR` circuit becomes persistent.

2. In order to re-encrypt the errors in the channel link, we should store the `obfuscators` in it too, and make them persistent too.

3. In current implementation switch as well as channel link creates the payment failures, so it means that we should store `obfuscator` in switch and also in channel links.

File: htlcswitch/switch.go (hunk 2 above)
Comment (AndrewSamokhvalov) [inline htlcswitch/switch.go]: insufficient capacity? 😛

File: htlcswitch/iterator.go [hunk 18]
```diff
@@ -185,3 +183,22 @@ func (p *SphinxDecoder) Decode(r io.Reader, rHash []byte) (HopIterator, error) {
 		processedPacket: sphinxPacket,
 	}, nil
 }
+
+// DecodeOnionObfuscator takes the onion blob as input extract the shard secret
```
Comment (Roasbeef) [inline htlcswitch/iterator.go]: shard -> shared

File: htlcswitch/link.go [hunk 19]
```diff
@@ -890,20 +905,30 @@ func (l *channelLink) processLockedInHtlcs(
...
+				case sphinx.ErrInvalidOnionVersion:
+					failure = lnwire.NewInvalidOnionVersion(onionBlob[:])
+				case sphinx.ErrInvalidOnionHMAC:
+					failure = lnwire.NewInvalidOnionHmac(onionBlob[:])
+				case sphinx.ErrInvalidOnionKey:
+					failure = lnwire.NewInvalidOnionKey(onionBlob[:])
+				default:
+					failure = lnwire.NewTemporaryChannelFailure(nil)
+				}
+				// If we're unable to parse the sphinx packet, then we'll cancel
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Missing a newline above this fragment.

File: htlcswitch/link.go [hunk 20]
```diff
@@ -1015,8 +1037,16 @@ func (l *channelLink) processLockedInHtlcs(
...
 
-					reason := []byte{byte(lnwire.IncorrectValue)}
-					l.sendHTLCError(pd.RHash, reason)
+					var failure lnwire.FailureMessage
+					update, err := l.createChannelUpdate()
+					if err != nil {
+						failure = lnwire.NewTemporaryChannelFailure(nil)
+					} else {
+						failure = lnwire.NewAmountBelowMinimum(
+							l.cfg.FwrdingPolicy.MinHTLC, *update)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: The value passed here is meant to be the value of the _incoming_ HTLC, not the value of the min HTLC policy.

File: htlcswitch/link.go [hunk 21]
```diff
@@ -1045,9 +1075,16 @@ func (l *channelLink) processLockedInHtlcs(
...
 
-					// TODO(andrew.shvv): send proper back error
-					reason := []byte{byte(lnwire.IncorrectValue)}
-					l.sendHTLCError(pd.RHash, reason)
+					var failure lnwire.FailureMessage
+					update, err := l.createChannelUpdate()
+					if err != nil {
+						failure = lnwire.NewTemporaryChannelFailure(nil)
+					} else {
+						failure = lnwire.NewFeeInsufficient(expectedFee,
```
Comment (Roasbeef) [inline htlcswitch/link.go]: The value here is meant to be the the value of the _incoming_ HTLC, not the fee we expect.

File: htlcswitch/link.go [hunk 22]
```diff
@@ -1061,16 +1098,22 @@ func (l *channelLink) processLockedInHtlcs(
...
-					l.sendHTLCError(pd.RHash, reason)
+					update, err := l.createChannelUpdate()
+					if err != nil {
+						l.fail("unable to create channel update "+
+							"while handling the error: %v", err)
+						return nil
+					}
+
+					failure := lnwire.NewIncorrectCltvExpiry(
+						fwdInfo.OutgoingCTLV, *update)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: The value here is meant to be value of the incoming _HTLC's_ timelock.

File: htlcswitch/switch.go [hunk 23]
```diff
@@ -300,11 +309,39 @@ func (s *Switch) handleLocalDispatch(payment *pendingPayment, packet *htlcPacket
...
 		// Retrieving the fail code from byte representation of error.
 		var userErr error
-		if code, err := htlc.Reason.ToFailCode(); err != nil {
-			userErr = errors.Errorf("can't decode fail code id"+
-				"(%v): %v", htlc.ID, err)
+
+		failure, err := payment.deobfuscator.Deobfuscate(htlc.Reason)
+		if err != nil {
+			userErr = errors.Errorf("unable to de-obfuscate "+
+				"onion failure, htlc with id(%v): %v", htlc.ID, err)
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: I'd say the payment hash of the HTLC is more relevant to a user than the ID (which they don't control or set).

File: htlcswitch/switch.go [hunk 24]
```diff
@@ -300,11 +309,39 @@ func (s *Switch) handleLocalDispatch(payment *pendingPayment, packet *htlcPacket
...
+			case *lnwire.FailExpiryTooSoon:
+				update = &failure.Update
+			case *lnwire.FailChannelDisabled:
+				update = &failure.Update
+			}
+
+			if update != nil {
+				log.Info("Received payment failure(%v), applying lightning "+
+					"network topology update", failure.Code())
+				s.cfg.UpdateTopology(update)
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: We could completely eliminate the requirement of the switch to _manually_ update the directed edge info when it encounters a payment error. Rather than up calling into the router to update the information, it could instead just return the error as is. However, it wouldn't be a typical error, instead the router would have a special `PaymentError` type. When a payment errors out, it can perform a type switch on the error. If it's this particular error, then it can decide exactly what to do with it based on the returned `lnwire.FailureCode`. Optionally an update would also be attached. 

The above method allows the router complete control over its actions when it receive an error. For example, if we send a payment and receive an error that says our fee is out of date, then can modify the route, and immediately re-send the payment. Other situations would be handled case-by-case. 

Note that this isn't a blocker for the PR, if you opt not to implement it, it can be done in some follow up work by either yourself or me.

File: htlcswitch/link.go [hunk 25]
```diff
@@ -410,40 +421,52 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket) {
...
+				if err != nil {
+					log.Errorf("unable to obfuscate error: %v", err)
+					return
+				}
+
+				go l.cfg.Switch.forward(newFailPacket(
+					l.ShortChanID(),
+					&lnwire.UpdateFailHTLC{
+						Reason: reason,
+					}, htlc.PaymentHash, htlc.Amount, false,
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Shouldn't the last parameter here be set to true? As the error has already been encrypted once.

File: htlcswitch/switch_test.go [hunk 26]
```diff
@@ -345,26 +363,30 @@ func TestSwitchSendPayment(t *testing.T) {
 		t.Fatal("wrong amount of circuits")
 	}
 
-	// Create fail request pretending that bob channel link handled the add
-	// htlc request with error  and sent the htlc fail request back. This
-	// request should be forwarder back to alice channel link.
-	packet := newFailPacket(
-		aliceChannelLink.ShortChanID(),
+	// Create fail request pretending that bob channel link handled
+	// the add htlc request with error  and sent the htlc fail request
```
Comment (Roasbeef) [inline htlcswitch/switch_test.go]: extra space after `error` here.

File: htlcswitch/switch_test.go [hunk 27]
```diff
@@ -345,26 +363,30 @@ func TestSwitchSendPayment(t *testing.T) {
...
+	// back. This request should be forwarder back to alice channel link.
```
Comment (Roasbeef) [inline htlcswitch/switch_test.go]: forwarder -> forwarded

File: lnwire/lnwire_test.go [hunk 28]
```diff
@@ -326,6 +326,23 @@ func TestLightningWireProtocol(t *testing.T) {
 
 			v[0] = reflect.ValueOf(*req)
 		},
+		MsgUpdateFailMalformedHTLC: func(v []reflect.Value, r *rand.Rand) {
```
Comment (Roasbeef) [inline lnwire/lnwire_test.go]: Ping on this comment.

File: lnwire/message.go [hunk 29]
```diff
@@ -107,12 +110,18 @@ func (u *UnknownMessage) Error() string {
 		u.messageType)
 }
 
+// Serializable is an interface which defines a lightning wire serializable
+// object.
+type Serializable interface {
+	Decode(io.Reader, uint32) error
```
Comment (Roasbeef) [inline lnwire/message.go]: Missing `godoc` comments on both of these methods.

File: lnwire/onion_error.go [hunk 30]
```diff
@@ -0,0 +1,978 @@
...
+
+// FailCode specifies the precise reason that an upstream HTLC was cancelled.
+// Each UpdateFailHTLC message carries a FailCode which is to be passed back
+// unaltered to the source of the HTLC within the route.
+type FailCode uint16
+
+// The currently defined onion failure types within this current version of the
+// Lightning protocol.
+const (
+	CodeInvalidRealm                  FailCode = FlagBadOnion | 1
```
Comment (Roasbeef) [inline lnwire/onion_error.go]: These are all missing `godoc` comments.

File: lnwire/onion_error.go [hunk 31]
```diff
@@ -0,0 +1,978 @@
...
+	return err
+}
+
+// FailAmountBelowMinimum is returned if the HTLC does not reach the
+// current minimum amount, we tell them the amount of the incoming HTLC
+// and the current channel setting for the outgoing channel.
+//
+// NOTE: might be return by forwarding node only.
+type FailAmountBelowMinimum struct {
+	HtlcMsat btcutil.Amount
```
Comment (Roasbeef) [inline lnwire/onion_error.go]: Missing a `godoc` comment.

File: lnwire/onion_error.go [hunk 32]
```diff
@@ -0,0 +1,978 @@
...
+	return f.Update.Encode(w, pver)
+}
+
+// FailFeeInsufficient is returned if the HTLC does not pay sufficient
+// fee, we tell them the amount of the incoming HTLC and the current
+// channel setting for the outgoing channel.
+//
+// NOTE: might be return by forwarding node only.
+type FailFeeInsufficient struct {
+	HtlcMsat btcutil.Amount
```
Comment (Roasbeef) [inline lnwire/onion_error.go]: Missing a `godoc` comment.

File: lnwire/onion_error.go [hunk 33]
```diff
@@ -0,0 +1,978 @@
...
+}
+
+// FailIncorrectCltvExpiry is returned if outgoing cltv value does not
+// match the update add htlc's cltv expiry minus cltv expiry delta
+// for the outgoing channel, we tell them the cltv expiry and the
+// current channel setting for the outgoing channel.
+//
+// NOTE: might be return by forwarding node only.
+type FailIncorrectCltvExpiry struct {
+	CltvExpiry uint32
```
Comment (Roasbeef) [inline lnwire/onion_error.go]: Missing a `godoc` comment.

File: lnwire/onion_error.go [hunk 34]
```diff
@@ -0,0 +1,978 @@
...
+	return f.Update.Encode(w, pver)
+}
+
+// FailFinalIncorrectCltvExpiry is returned if the outgoing_cltv_value does not
+// match the ctlv_expiry of the HTLC at the final hop.
+//
+// NOTE: might be returned by final node only.
+type FailFinalIncorrectCltvExpiry struct {
+	// CltvExpiry
+	CltvExpiry uint32
```
Comment (Roasbeef) [inline lnwire/onion_error.go]: The comment should be completed.

File: lnwire/onion_error.go [hunk 35]
```diff
@@ -0,0 +1,978 @@
...
+func (f *FailFinalIncorrectCltvExpiry) Encode(w io.Writer, pver uint32) error {
+	return writeElement(w, f.CltvExpiry)
+}
+
+// FailFinalIncorrectHtlcAmount is returned if the amt_to_forward is
+// higher than incoming_htlc_amt of the HTLC at the final hop.
+//
+// NOTE: might be returned by final node only.
+type FailFinalIncorrectHtlcAmount struct {
+	IncomingHTLCAmount btcutil.Amount
```
Comment (Roasbeef) [inline lnwire/onion_error.go]: Missing a `godoc` comment.

File: peer.go [hunk 36]
```diff
@@ -306,11 +306,34 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
 		// new payments triggered by RPC clients.
-		sphinxDecoder := htlcswitch.NewSphinxDecoder(p.server.sphinx)
 		link := htlcswitch.NewChannelLink(
 			htlcswitch.ChannelLinkConfig{
-				Peer:             p,
-				DecodeOnion:      sphinxDecoder.Decode,
+				Peer:                  p,
+				DecodeHopIterator:     p.server.sphinx.DecodeHopIterator,
+				DecodeOnionObfuscator: p.server.sphinx.DecodeOnionObfuscator,
+				ChannelEdgeInfo: func(chanID lnwire.ChannelID) (*channeldb.ChannelEdgePolicy,
```
Comment (Roasbeef) [inline peer.go]: As the switch only needs to obtain the latest channel update from _us_, we can simplify this function and the handling within the switch a good bit. Instead of accepting the `chanID` here, we'll just thread a function through that only return the latest channel update. This will let us not have the logic of parsing the feature bits of an update to decide which direction should be returned.

File: peer.go [hunk 37]
```diff
@@ -784,11 +806,34 @@ out:
...
 
-			decoder := htlcswitch.NewSphinxDecoder(p.server.sphinx)
 			link := htlcswitch.NewChannelLink(
 				htlcswitch.ChannelLinkConfig{
-					Peer:             p,
-					DecodeOnion:      decoder.Decode,
+					Peer:                  p,
+					DecodeHopIterator:     p.server.sphinx.DecodeHopIterator,
+					DecodeOnionObfuscator: p.server.sphinx.DecodeOnionObfuscator,
+					ChannelEdgeInfo: func(chanID lnwire.ChannelID) (*channeldb.ChannelEdgePolicy,
```
Comment (Roasbeef) [inline peer.go]: Similar comment here. 

Also as this logic is repeated above, a function should be added which factors out the logic. Note that the function need not have a struct as its target (a method), it can be free form.

File: htlcswitch/switch.go (hunk 24 above)
Comment (AndrewSamokhvalov) [inline htlcswitch/switch.go]: I would prefer to do that in another `PR` just because when I started to changing the way how we return the error, I realised that we a mixing local and remote errors, for that reason we can't just swap the `error` to the `lnwire.FailureMessage`, and it would require a bit of the thinking. Would like to do that after `PR` will be merged.

File: peer.go [hunk 38]
```diff
@@ -1363,3 +1367,51 @@ func (p *peer) PubKey() [33]byte {
...
+		cb := func(chanInfo *channeldb.ChannelEdgeInfo,
+			e1, e2 *channeldb.ChannelEdgePolicy) error {
+			if lnwire.NewChanIDFromOutPoint(&chanInfo.ChannelPoint) == chanID {
+				edge1 = e1
+				edge2 = e2
+			}
+			return nil
+		}
+
+		err := router.ForEachChannel(cb)
```
Comment (Roasbeef) [inline peer.go]: This method will iterate over _all_ channels within the _entire_ graph. A quicker method would be to use the `GenPossibleOutpoints` method. Although this may in the worse case generate 65k candidates, in practice, the outpoint index value will be between 1 and 10, so this should exit rather quickly.

File: peer.go [hunk 39]
```diff
@@ -306,11 +307,13 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
-		sphinxDecoder := htlcswitch.NewSphinxDecoder(p.server.sphinx)
 		link := htlcswitch.NewChannelLink(
 			htlcswitch.ChannelLinkConfig{
-				Peer:             p,
-				DecodeOnion:      sphinxDecoder.Decode,
+				Peer:                  p,
+				DecodeHopIterator:     p.server.sphinx.DecodeHopIterator,
+				DecodeOnionObfuscator: p.server.sphinx.DecodeOnionObfuscator,
+				GetLastChannelUpdate: createGetLastUpdate(p.server.chanRouter,
+					p.PubKey(), chanID),
```
Comment (Roasbeef) [inline peer.go]: We have access to the channel point here above, so we can pass that in directly. If we do this, this let's us avoid looping over all the edges in the entire graph each time we need to send an error which requires an update.

File: peer.go (hunk 38 above)
Comment (AndrewSamokhvalov) [inline peer.go]: Probably would be better to use the `GetChannelByID`

File: lnwire/onion_error.go [hunk 40]
```diff
@@ -59,9 +61,12 @@ const (
 	CodeChannelDisabled                        = FlagUpdate | 20
 	CodeUnknownPaymentHash                     = FlagPerm | 15
 	CodeIncorrectPaymentAmount                 = FlagPerm | 16
-	CodeFinalExpiryTooSoon                     = 17
-	CodeFinalIncorrectCltvExpiry               = 18
-	CodeFinalIncorrectHtlcAmount               = 19
+	CodeFinalExpiryTooSoon            FailCode = 17
```
Comment (AndrewSamokhvalov) [inline lnwire/onion_error.go]: https://play.golang.org/p/cxHOZ4LpcB
